src/lib/CofThy.ml

```diff
@@ -1,209 +1,278 @@
-    cont @@ `Consistent alg_thy'
+  let init : t = [Alg.init', (VarSet.empty, [])]
+
+  let consistency =
+    function
+    | [] -> `Inconsistent
+    | _ -> `Consistent
+
+  (** [refactor_branches] attempts to identify common parts of the branches
+    * and shrink the labels. Recall that we do not keep the common ancestor
+    * but the paths (as a collection of atomic cofibrations) from the common
+    * ancestor, and thus what are changed here are the paths.
+    *
+    * This optimization seems to be expensive, but it seems to help after
+    * we switched from the persistant tables (using [Hashtbl]) to [Map].
+  *)
+  let refactor_branches cached_branches : t =
+    let common_vars =
+      let go vars0 (_, (vars1, _)) = VarSet.inter vars0 vars1 in
+      match cached_branches with
+      | [] -> VarSet.empty
+      | (_, (vars, _)) :: branches -> List.fold_left go vars branches
+    in
+    (* The following is an approximation of checking whether some equation is useful.
+     * It does not kill every "useless" cofibration. Here is one example:
+     *
+     * branch 1: r=0
+     * branch 2: r=i, i=0
+     *
+     * r=0 will be factored out, but then a=0 should also be removed. Here is a more
+     * complicated example:
+     *
+     * branch 1: r=i, i=0
+     * branch 2: r=j, j=0
+     *
+     * Both i=0 and j=0 should be factored out, but the following code is not smart
+     * enough to detect them. One could consider more aggressive approaches if [CofThy]
+     * becomes the bottleneck again.
+    *)
+    let useful eq = cached_branches |> List.exists @@ fun (thy', _) -> not @@ Alg.test_eq thy' eq in
+    (* revisit all branches and remove all useless ones identified by the simple criterion above. *)
+    cached_branches |> List.map @@ fun (thy', (vars, eqs)) ->
+    thy', (VarSet.diff vars common_vars, List.filter useful eqs)
+
+  (** [split] incorporate a cofibration context and split an already split theory
+    * into more branches. This is similar to [Alg.split] in the spirit, but differs
+    * in lots of details. *)
+  let split (thy : t) (cofs : cof list) : t =
+    match dissect_cofibrations cofs with
+    | [] -> []
+    | [vars, []] when VarSet.is_empty vars -> thy
+    | dissected_cofs ->
+      Alg.drop_useless_branches begin
+        thy |> List.concat_map @@ fun (thy', (vars, eq)) ->
+        Alg.reduce_branches thy' dissected_cofs |> List.map @@ fun (thy', (sub_vars, sub_eqs)) ->
+        thy', (VarSet.union vars sub_vars, eq @ sub_eqs)
+      end
+
+  (** [assume] is the same as [split] except that it will refactor the branches. *)
+  let assume (thy : t) (cofs : cof list) : t =
+    refactor_branches @@ split thy cofs
+
+  let test_sequent thy cx cof =
+    begin
+      split thy cx |> List.map (fun (thy', _) -> thy')
+    end |> List.for_all @@ fun thy' -> Alg.test thy' cof
+
+  let left_invert ~zero ~seq thy cont =
```

Comment:
I am not familiar enough anymore with the algorithm you have implemented to say whether this is still left inversion. Is it still left inversion, or should it be renamed something like `search`?

Reply:
In case it helps, the disjunctive (normal?) form is implementing eager left inversion. That is, it precomputes all the branches during left inversion the moment any cofibration is added to the theory. Please rename the function to whatever you like. (Do you still like `search` with this explanation?)

Reply:
Gotcha; so it sounds like you are computing the search tree when things are added, and then this function is just traversing the search tree. Is that right?

Reply:
Yes. All the optimizers are to prune the precomputed trees.

Reply:
PS: I kept the name only because I did not change its semantics.